src/main/java/com/example/forum/member/service/impl/MemberServiceImpl.java

```diff
@@ -0,0 +1,132 @@
+
+@RequiredArgsConstructor
+@Service
+public class MemberServiceImpl implements MemberService {
+
+
+  private final MemberRepository memberRepository;
+  private final EmailComponent emailComponent;
+  private final PasswordEncoder passwordEncoder;
+  private final AuthenticationManager authenticationManager;
+
+  private final CustomUserDetailsService customUserDetailsService;
+
+
+  /**
+   * 회원가입 - 중복된 이메일 가입 불가 - 이메일 인증키 생성 - 인증 이메일 발송
+   */
+  @Override
+  public void signUp(MemberInput memberInput) {
+
+    // duplication check
+    if (memberRepository.countByEmail(memberInput.getEmail()) > 0) {
+      throw new ExistEmailException("이미 존재하는 이메일입니다.");
+    }
+
+    // member 생성
+    // password 암호화
+    String encodedPassword = passwordEncoder.encode(memberInput.getPassword());
+    // authentication key 값 생성
+    String uuid = UUID.randomUUID().toString();
+
+    Member member = new Member().builder()
+        .email(memberInput.getEmail())
+        .password(encodedPassword)
+        .username(memberInput.getUsername())
+        .isAuthenticated(false)
+        .emailAuthKey(uuid)
+        .registeredDate(LocalDateTime.now())
+        .role(Role.USER)
+        .build();
+
+    memberRepository.save(member);
+
+    // send email authentication key
+    String email = member.getEmail();
+    String subject = "forum 가입을 축하합니다.";
+    String text = "<p>사이트 가입을 축하합니다</p> <p>아래 링크를 클릭하셔서 가입을 완료하세요.</p> " +
+        "<div><a href='http://localhost:8080/member/email-auth?id=" + uuid + "'> 가입 완료</a></div>";
+
+    emailComponent.sendEmail(email, subject, text);
+  }
+
+
+  /**
+   * 이메일 인증
+   */
+  public ResponseEntity<?> authenticateEmail(String id) {
+
+    Optional<Member> member = memberRepository.findByEmailAuthKey(id);
+
+    if (member.isPresent()) {
+
+      member.get().setAuthenticated(true);
+      memberRepository.save(member.get());
+
+      return new ResponseEntity<>("Email 인증 성공.", HttpStatus.OK);
+
+    } else {
+      return new ResponseEntity<>("Email 인증 실패.", HttpStatus.BAD_REQUEST);
+    }
+
+  }
+
+  /**
+   * 로그인
+   */
+
+  public ResponseEntity<?> login(LoginInput loginInput) {
+    UserDetails userDetails = customUserDetailsService.loadUserByUsername(loginInput.getEmail());
+
+    try {
+      authenticationManager.authenticate(
+          new UsernamePasswordAuthenticationToken(loginInput.getEmail(), loginInput.getPassword()));
+
+      SecurityContextHolder.getContext().setAuthentication(
+          new UsernamePasswordAuthenticationToken(userDetails, null, userDetails.getAuthorities()));
+
+      return ResponseEntity.ok("로그인 성공!");
+    } catch (AuthenticationException e) {
+      return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body("로그인 실패");
+    }
+  }
+
+  /**
+   * 로그아웃
+   */
+  public ResponseEntity<?> logout(HttpServletRequest request) {
+    Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+
+    if(authentication != null){
```

Comment:
```suggestion
    if (authentication != null) {
```
coding convention 에 맞지않는 style 인데요. 꼭 맞추도록 노력해보시면 좋겠네요~!
https://naver.github.io/hackday-conventions-java/#space-around-brace